Additions to persistence PR

BenjaminBossan opened the pull request:
- Slight change in signature: `save(obj, file)` instead of
  `save(file, obj)`, as it's more consistent with existing saving functions
- Rename `cls` to `cls_` in `get_instance_method` to prevent variable name
  reuse
- Add `Persister` class that allows users to add their custom methods for
  getting and loading objects.
- Add support for `Pipeline` and `FeatureUnion`
- Add support for basic numpy functions
- Add test scaffold + a couple of tests (some of them `xfail`ing for now)

BenjaminBossan committed:
Additions to persistence PR

- Slight change in signature: save(obj, file) instead of save(file,
  obj), as it's more consistent with existing saving functions
- Rename cls to cls_ in get_instance_method to prevent variable name
  reuse
- Add Persister class that allows users to add their custom methods for
  getting and loading objects.
- Add support for Pipeline and FeatureUnion
- Add support for basic numpy functions
- Add test scaffold + a couple of tests (some of them xfailing for now)

BenjaminBossan commented:
The discussed changes @adrinjalali

adrinjalali commented on skops/_persist.py:
```diff
@@ -1,24 +1,37 @@
 from sklearn.base import BaseEstimator
+from sklearn.pipeline import FeatureUnion, Pipeline
+
+import skops
```
why do we need this?

adrinjalali commented on skops/_persist.py:
```diff
@@ -1,24 +1,37 @@
 
 
-def BaseEstimator_get_state(obj, dst):
+def BaseEstimator_get_state(obj, dst, methods):
```
I would hold on with adding the extra method dispatch thing and think about that in a later PR, there are other ways to do it as well I think.

adrinjalali commented on skops/_persist.py:
if we hold on for the extra methods thing here, then we also don't need this class for now.

adrinjalali commented on skops/test_persist.py:
```diff
@@ -0,0 +1,378 @@
+            assert val0 == val1
+
+
+class EstimatorBaseTester:
```
For common tests, I rather do the way I now put in the PR: https://github.com/skops-dev/skops/pull/128/files#diff-9ce8f30521c2b9fb6a1b2f0813e522810362fa16cc9a8be1660c4f5637dbcce1

It reduces the boilerplate code I think.

adrinjalali commented on skops/test_persist.py:
```diff
@@ -0,0 +1,378 @@
+        np.testing.assert_allclose(X_original, X_loaded)
+
+
+class TestClfPipelineAndFeatureUnion(EstimatorBaseTester):
```
happy to keep this one though, for a more in depth test here.

adrinjalali reviewed (commented):
Thanks @BenjaminBossan , this is quite a bit of work already done!

adrinjalali commented:
I also put the test under `skops/tests` instead of `skops/`

BenjaminBossan committed:
Address reviewer feedback

- Remove option to modify get and set method dicts
- Remove tests in favor of tests added by reviewer

BenjaminBossan commented:
@adrinjalali I made the requested changes. Some comments:

I can't seem to be able to fetch your latest changes, so after my merge all the tests you added will break since I changed the signature of `save`.

As for the testing approach, I was thinking about using the existing sklearn utilities to add tests but decided against it because:

1. They rely on private functions.
2. Most of the tested objects still fail - is there an easy way to mark a subset of them to be expected to fail? I don't know any, hence I wanted to opt them in explicitly.
3. Some special cases are not tested there, e.g. the deeply nested pipeline/feature union, `FunctionTransformer` with specific function types, etc., though that should not be too hard to add.

It's still okay to go that way but we need a solution for 2.

adrinjalali commented:
Yes, I agree we should do the more complicated tests as well, that's what I meant by this one: https://github.com/adrinjalali/skops/pull/1#discussion_r964611730

We'd have added tests specific to certain classes.

As for the failing tests, this is what we do as it's usually done in sklearn, with explicit ignore lists: https://github.com/scikit-learn/scikit-learn/blob/670133dbc42ebd9f79552984316bc2fcfd208e2e/sklearn/tests/test_docstrings.py#L14

As for using private methods, I agree, but rather still use them and work upstream to make them public or a part of the developer API. We can make sure things work for future releases by adding a test against the nightly builds.

adrinjalali commented:
This looks good, I'll merge and fix the rest on the PR.

BenjaminBossan commented:
> As for the failing tests, this is what we do as it's usually done in sklearn, with explicit ignore lists: https://github.com/scikit-learn/scikit-learn/blob/670133dbc42ebd9f79552984316bc2fcfd208e2e/sklearn/tests/test_docstrings.py#L14

I think for now we would have an easier time with opting in instead of opting out :D
